bundles/org.eclipse.e4.ui.workbench.renderers.swt/src/org/eclipse/e4/ui/workbench/renderers/swt/StackRenderer.java

```diff
@@ -209,6 +218,32 @@ private void addItemToSet(List<CTabItem> itemsToSet, MPart parentParent) {
 		}
 	}
 
+	@Inject
+	@Optional
+	void subscribePerspectiveSwitched(
+			@UIEventTopic(UIEvents.UILifeCycle.PERSPECTIVE_SWITCHED) org.osgi.service.event.Event event) {
+		Object element = event.getProperty(EventTags.ELEMENT);
+		if (element instanceof MPerspective) {
+			setOnboarding((MPerspective) element);
+		}
+	}
+
+	private void setOnboarding(MPerspective perspective) {
+		welcomeText.getDisplay().asyncExec(() -> {
```

Comment:
This will fail with NPE if there was no "org.eclipse.e4.primaryDataStack" tag at the "createWidget()" time.

Reply:
fixed